File: Python/002.py
def hantei(S):
    cnt = 0
    for i in range(len(S)):
        if S[i] == '(':
            cnt += 1  
        else:
            cnt -= 1   
        if cnt < 0:
            return False 
    if cnt == 0:
        return True
    else:
        return False

N = int(input())

for i in range(1 << N):             #1<<Nは2^Nという意味
    candidate = ""
    for j in reversed(range(N)):        #reversedした理由は左から確認していくから（右から確認するときはreversedなし）
        if (i & (1 << j)) == 0:
            candidate += "("
        else:
            candidate += ")"
    h = hantei(candidate)
    if h == True:
        print(candidate)

File: C++/002.cpp
#include <bits/stdc++.h>
using namespace std;

bool hantei(string S) {
    int dep = 0;
    for (int i = 0; i<S.size(); i++) {
        if (S[i] == '(') dep += 1;
        if (S[i] == ')') dep -= 1;
        if (dep < 0) return false;
    }
    if (dep == 0) return true;
    return false;
}

int main() {
    int N;
    cin >> N;
    for (int i = 0; i < (1 << N); i++) {        //1<<Nは2^Nという意味（2**Nと同じ意味)
        string candidate = "";
        for (int j = N - 1; j >= 0; j--) {
            if ((i & (1 << j)) == 0) {              //左からbitが立っているか立っていないかをチェック
                candidate += "(";                   //0のとき '('
            }
            else {
                candidate += ")";                   //1のとき ')'
            }
        }
        bool I = hantei(candidate);
        if (I == true) cout << candidate << endl;
    }
}